plutus-core/plutus-ir/src/PlutusIR/Transform/LetFloatIn.hs

```diff
@@ -0,0 +1,210 @@
+        Let a Rec bs0 body0 ->
+            -- Currently we don't move recursive bindings, so we simply descend into the body.
+            let bs = goBinding <$> bs0
+                body = go body0
+                us = Set.union (termUniqs body) (foldMap bindingUniqs bs)
+             in Let (a, us) Rec bs body
+        Var a n -> Var (a, Set.singleton (n ^. PLC.unique)) n
+        Constant{} -> (,mempty) <$> t
+        Builtin{} -> (,mempty) <$> t
+        Error{} -> (,mempty) <$> t
+
+    -- | Float bindings in the given `Binding` inwards, and calculate the set of
+    -- variable `Unique`s in the result `Binding`.
+    goBinding ::
+        Binding tyname name uni fun a ->
+        Binding tyname name uni fun (a, Uniques)
+    goBinding = \case
+        TermBind a s var rhs ->
+            let rhs' = go rhs
+             in TermBind (a, termUniqs rhs') s (noUniq var) rhs'
+        other -> noUniq other
+
+termUniqs :: Term tyname name uni fun (a, Uniques) -> Uniques
+termUniqs = snd . termAnn
+
+-- | The result includes both the `Unique` of the bound var and the `Unique`s in the RHS.
+bindingUniqs ::
+    PLC.HasUnique name PLC.TermUnique =>
+    Binding tyname name uni fun (a, Uniques) ->
+    Uniques
+bindingUniqs b = Set.union (snd (bindingAnn b)) $ case b of
+    TermBind _a _s var _rhs -> Set.singleton $ var ^. PLC.varDeclName . PLC.unique
+    _                       -> mempty
+
+noUniq :: Functor f => f a -> f (a, Uniques)
+noUniq = fmap (,mempty)
+
+{- | Given a `Term` and a `Binding`, determine whether the `Binding` can be
+ placed somewhere inside the `Term`.
+
+ If yes, return the result `Term`. Otherwise, return a `Let` constructed from
+ the given `Binding` and `Term`.
+-}
+floatInBinding ::
+    forall tyname name uni fun a.
+    ( PLC.HasUnique name PLC.TermUnique
+    , PLC.ToBuiltinMeaning uni fun
+    ) =>
+    PLC.BuiltinVersion fun ->
+    -- | Annotation to be attached to the constructed `Let`.
+    a ->
+    Binding tyname name uni fun (a, Uniques) ->
+    Term tyname name uni fun (a, Uniques) ->
+    Term tyname name uni fun (a, Uniques)
+floatInBinding ver letAnn = go
+  where
+    go ::
+        Binding tyname name uni fun (a, Uniques) ->
+        Term tyname name uni fun (a, Uniques) ->
+        Term tyname name uni fun (a, Uniques)
+    go b !body = case b of
+        TermBind _a Strict _var rhs
+            -- A strict binding whose RHS is not pure cannot be moved.
+            | not (isPure ver (const NonStrict) rhs) ->
+                let us = Set.union (termUniqs body) (bindingUniqs b)
+                 in Let (letAnn, us) NonRec (pure b) body
+        TermBind (_, usBind) _strictness var _rhs -> case body of
+            Apply (a, usBody) fun arg
+                | (var ^. PLC.unique) `Set.notMember` termUniqs fun ->
+                    -- `fun` does not mention `var`, so we can place the binding inside `arg`.
+                    Apply (a, Set.union usBind usBody) fun (go b arg)
```

Comment:
This is a little subtle and it took me a while to realise exactly why we need this union: it's so that we consider the variables used by the _right-hand side_ of the binding to be used in this branch of the application. I'd really like a test case that demonstrates that we handle this properly.